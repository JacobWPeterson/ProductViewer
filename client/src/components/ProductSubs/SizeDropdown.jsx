import React, { useEffect, useState } from 'react';
import styled from 'styled-components';
import FontAwesome from 'react-fontawesome';

const DDWrapper = styled.div`
  position: relative;
  width: 65%;
  font-size: 0;
  user-select: none;
`;

const DDHeader = styled.button`
  font-size: 1rem;
  position: relative;
  height: 7vh;
  width: 100%;
  padding: 0 1.25vw;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: white;
  color: #424242;
  border: 1px solid #424242;
  cursor: default;
  cursor: pointer;
  &:focus {
    outline: none;
    z-index: 1;
    border-top: 1px solid #e0e0e0;
    border-right: 1px solid #fff;
    border-left: 1px solid #fff;
    border-bottom: 1px solid #aeaeae;
    box-shadow: 0 3px 6px #a0a0a0, 0 3px 6px #a0a0a0;
  };
`;

const DDHeaderTitle = styled.div`
  font-weight: 600;
`;

const DDList = styled.div`
  position: absolute;
  z-index: 2;
  width: 100%;
  max-height: 25vh;
  overflow-y: scroll;
  overflow-x: hidden;
  font-weight: 400;
  box-shadow: 0 3px 6px #a0a0a0, 0 3px 6px #a0a0a0;
`;

const DDListItem = styled.button`
  display: inline-block;
  overflow: hidden;
  width: 100%;
  padding: 8px 10px;
  font-weight: 400;
  color: #424242;
  font-size: 1rem;
  cursor: default;
  cursor: pointer;
  border-width: 0;
  border-style: none;
  border-color: clear;
  background-color: white;
  border: 1px solid clear;
  border-bottom: 1px solid #aeaeae;
  &:hover {
    border-bottom: 2px solid #424242;
    border-top: 1px solid #aeaeae;
    transform: scale(1.1);
  };
`;

const SizeDropdown = (props) => {
  const {
    list, resetThenSet, title,
  } = props;
  const [isListOpen, setListStatus] = useState(false);
  const [headerTitle, setHeaderTitle] = useState(title);

  const close = () => {
    setListStatus(false);
  };

  useEffect(() => {
    if (isListOpen) {
      window.addEventListener('click', close);
      return () => window.removeEventListener('click', close);
    }
  });

  const toggleList = () => {
    setListStatus(!isListOpen);
  };

  const selectItem = (item) => {
    setHeaderTitle(item.size);
    setListStatus(false);
    resetThenSet(item.quantity, item.size);
  };

  return (
    <DDWrapper>

      <DDHeader onClick={toggleList}>
        <DDHeaderTitle>{headerTitle}</DDHeaderTitle>
        {isListOpen
          ? <FontAwesome name="angle-up" size="2x" />
          : <FontAwesome name="angle-down" size="2x" />}
      </DDHeader>

      {isListOpen && (
      <DDList role="list">
        {list.map((item) => (
          item.quantity > 0 && (
          <DDListItem type="button" key={item.size} onClick={() => selectItem(item)}>
            {item.size}
          </DDListItem>
          )
        ))}
      </DDList>
      )}

    </DDWrapper>
  );
};

export default SizeDropdown;
