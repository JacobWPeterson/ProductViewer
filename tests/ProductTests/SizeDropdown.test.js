/* eslint-disable no-undef */
/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import { mount } from 'enzyme';
import styled from 'styled-components';

import SizeDropdown from '../../client/src/components/ProductSubs/SizeDropdown.jsx';
import ProductTestData from '../ProductTestData.js';

const DDWrapper = styled.div`
  position: relative;
  width: 35%;
  font-size: 0;
  user-select: none;
  margin-left: 1.5vw;
`;

const DDHeader = styled.button`
  font-size: 1rem;
  position: relative;
  height: 7vh;
  width: 100%;
  padding: 0 1.25vw;
  display: flex;
  justify-content: space-between;
  align-items: center;
  background-color: white;
  color: #424242;
  border: 1px solid #424242;
  cursor: default;
  cursor: pointer;
  &:focus {
    outline: none;
    z-index: 1;
    border-top: 1px solid #e0e0e0;
    border-right: 1px solid #fff;
    border-left: 1px solid #fff;
    border-bottom: 1px solid #aeaeae;
    box-shadow: 0 3px 6px #a0a0a0, 0 3px 6px #a0a0a0;
  };
`;

const DDListItem = styled.button`
  display: inline-block;
  overflow: hidden;
  width: 100%;
  padding: 8px 10px;
  font-weight: 400;
  color: #424242;
  font-size: 1rem;
  cursor: default;
  cursor: pointer;
  border-width: 0;
  border-style: none;
  border-color: clear;
  background-color: white;
  border: 1px solid clear;
  border-bottom: 1px solid #aeaeae;
  &:hover {
    border-bottom: 2px solid #424242;
    border-top: 1px solid #aeaeae;
    transform: scale(1.1);
  };
`;

const props = {
  list: Object.values(ProductTestData.productStyles.results[0].skus),
};

describe('SizeDropdown', () => {
  const SizeDropdownComponent = mount(<SizeDropdown {...props} debug />);

  it('Should render correctly', () => {
    expect(SizeDropdownComponent).toMatchSnapshot();
  });

  const button = SizeDropdownComponent.find('button');
  // const ListItem = SizeDropdownComponent.find(DDListItem).at(6).text();

  // it('renders the correct list of available stock', () => {
  //   expect(ListItem).toEqual(props.available);
  // });

  it('toggles the list on dropdown header click', () => {
    button.simulate('click');
    expect(SizeDropdownComponent.state('isListOpen')).toEqual(true);
  });

  // Set-up event listener mock
  // window.addEventListener = jest.fn((event, callback) => {
  //   map[event] = callback;
  // });

  // it('toggles the list closed', () => {
  //   window.simulate('click');
  //   expect(SizeDropdownComponent.state('isListOpen')).toEqual(false);
  // });
});
