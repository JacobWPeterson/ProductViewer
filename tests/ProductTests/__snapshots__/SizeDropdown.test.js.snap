// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SizeDropdown Should render correctly 1`] = `
<SizeDropdown
  debug={true}
  list={
    Array [
      Object {
        "quantity": 8,
        "size": "XS",
      },
      Object {
        "quantity": 16,
        "size": "S",
      },
      Object {
        "quantity": 17,
        "size": "M",
      },
      Object {
        "quantity": 10,
        "size": "L",
      },
      Object {
        "quantity": 15,
        "size": "XL",
      },
      Object {
        "quantity": 6,
        "size": "XXL",
      },
    ]
  }
>
  <SizeDropdown__DDWrapper>
    <div
      className="SizeDropdown__DDWrapper-u3ohyw-0 knauff"
    >
      <SizeDropdown__DDHeader
        onClick={[Function]}
      >
        <button
          className="SizeDropdown__DDHeader-u3ohyw-1 dFNLhV"
          onClick={[Function]}
        >
          <SizeDropdown__DDHeaderTitle>
            <div
              className="SizeDropdown__DDHeaderTitle-u3ohyw-2 fzssNQ"
            />
          </SizeDropdown__DDHeaderTitle>
          <FontAwesome
            name="angle-down"
            size="2x"
          >
            <span
              aria-hidden={true}
              className="fa fa-angle-down fa-2x"
            />
          </FontAwesome>
        </button>
      </SizeDropdown__DDHeader>
    </div>
  </SizeDropdown__DDWrapper>
</SizeDropdown>
`;
